lib/pub_grub/version_solver.rb

```diff
@@ -104,25 +107,15 @@ def propagate_incompatibility(incompatibility)
       unsatisfied.package
     end
 
-    def next_term_to_try_from(unsatisfied_terms)
-      unsatisfied_terms.min_by do |term|
-        package = term.package
-        range = term.constraint.range
-        matching_versions = source.versions_for(package, range)
-        higher_versions = source.versions_for(package, range.upper_invert)
-
-        [matching_versions.count <= 1 ? 0 : 1, higher_versions.count]
-      end
-    end
-
     def choose_package_version_from(unsatisfied_terms)
-      unsatisfied_term = next_term_to_try_from(unsatisfied_terms)
-      package = unsatisfied_term.package
+      remaining = unsatisfied_terms.map { |t| [t.package, t.constraint.range] }.to_h
+
+      package, version = strategy.next_package_and_version(remaining)
 
-      version = source.versions_for(package, unsatisfied_term.constraint.range).first
       logger.debug { "attempting #{package} #{version}" }
 
       if version.nil?
+        unsatisfied_term = unsatisfied_terms.find { |t| t.package == package }
```

Comment:
This line is also a bit weird. Since we no longer operate on the term (only package/version), we have to re-lookup the term to create the incompatibility.